src/Microsoft.AspNetCore.Authentication.OpenIdConnect/Events/AuthorizationCodeReceivedContext.cs

```diff
@@ -17,29 +18,70 @@ public class AuthorizationCodeReceivedContext : BaseOpenIdConnectContext
-        /// Gets or sets the 'code'.
+        /// Gets or sets the <see cref="JwtSecurityToken"/> that was received in the authentication response, if any.
         /// </summary>
-        public string Code { get; set; }
+        public JwtSecurityToken JwtSecurityToken { get; set; }
 
         /// <summary>
-        /// Gets or sets the <see cref="JwtSecurityToken"/> that was received in the id_token + code OpenIdConnectRequest.
+        /// The request that will be sent to the token endpoint and is available for customization.
         /// </summary>
-        public JwtSecurityToken JwtSecurityToken { get; set; }
+        public OpenIdConnectMessage TokenEndpointRequest { get; set; }
+
+        /// <summary>
+        /// The configured communication channel to the identity provider for use when making custom requests to the token endpoint.
+        /// </summary>
+        public HttpClient Backchannel { get; internal set; }
+
+        /// <summary>
+        /// If the developer chooses to redeem the code themselves then they can provide the resulting tokens here. This is the
+        /// same as calling SkipCodeRedemption. If set then the middleware will not attempt to redeem the code. An IdToken
+        /// is required if one had not been previously received in the authorization response. An access token is optional
+        /// if the middleware is to contact the user-info endpoint.
+        /// </summary>
+        public OpenIdConnectMessage TokenEndpointResponse { get; set; }
+
+        /// <summary>
+        /// Tells the middleware to skip the code redemption process. The developer may have redeemed the code themselves, or
+        /// decided that the redemption was not required. If tokens were retrieved that are needed for further processing then
+        /// call one of the overloads that allows providing tokens. An IdToken is required if one had not been previously received
+        /// in the authorization response. An access token can optionally be provided for the middleware to contact the
+        /// user-info endpoint. Calling this is the same as setting TokenEndpointResponse.
+        /// </summary>
+        public void SkipCodeRedemption()
+        {
+            TokenEndpointResponse = new OpenIdConnectMessage();
+        }
+
+        /// <summary>
+        /// Tells the middleware to skip the code redemption process. The developer may have redeemed the code themselves, or
+        /// decided that the redemption was not required. If tokens were retrieved that are needed for further processing then
+        /// call one of the overloads that allows providing tokens. An IdToken is required if one had not been previously received
+        /// in the authorization response. An access token can optionally be provided for the middleware to contact the
+        /// user-info endpoint. Calling this is the same as setting TokenEndpointResponse.
+        /// </summary>
+        public void SkipCodeRedemption(string accessToken)
```

Comment:
Not sure having such an overload is really pertinent as it doesn't allow setting the identity token, which MUST be provided by the authorization server in the token response (in both the code and the implicit flows).